include/seqan3/alignment/matrix/debug_matrix.hpp

```diff
@@ -0,0 +1,394 @@
+     * \param cols    The number of columns.
+     * \param sequence1 The first sequence of the sequence alignment.
+     * \param sequence2 The second sequence of the sequence alignment.
+     */
+    debug_matrix(std::vector<entry_type> entries, size_t const rows, size_t const cols,
+                 sequence1_t sequence1, sequence2_t sequence2)
+        : _matrix{std::move(entries), rows, cols},
+          _sequence1{std::forward<sequence1_t>(sequence1)},
+          _sequence2{std::forward<sequence2_t>(sequence2)}
+    {
+        assert(cols == _sequence1.size() + 1u);
+        assert(rows == _sequence2.size() + 1u);
+    }
+
+    /*!\brief Construct the matrix out of an existing matrix.
+     * \param matrix An alignment matrix; Must model seqan3::detail::Matrix.
+     */
+    debug_matrix(matrix_t matrix)
+        : _matrix{std::forward<matrix_t>(matrix)},
+          _sequence1{std::nullopt},
+          _sequence2{std::nullopt}
+    {}
+
+    /*!\brief Construct the matrix out of an existing matrix and two sequences.
+     * \param matrix An alignment matrix; Must model seqan3::detail::Matrix.
+     * \param sequence1 The first sequence of the sequence alignment.
+     * \param sequence2 The second sequence of the sequence alignment.
+     */
+    debug_matrix(matrix_t matrix, sequence1_t sequence1, sequence2_t sequence2)
+        : _matrix{std::forward<matrix_t>(matrix)},
+          _sequence1{std::forward<sequence1_t>(sequence1)},
+          _sequence2{std::forward<sequence2_t>(sequence2)}
+    {
+        assert(_matrix.cols() == _sequence1.size() + 1u);
+        assert(_matrix.rows() == _sequence2.size() + 1u);
+    }
+    //!\}
+
+    //!\copydoc seqan3::detail::Matrix::rows
+    size_t rows() const noexcept
+    {
+        return _matrix.rows();
+    }
+
+    //!\copydoc seqan3::detail::Matrix::cols
+    size_t cols() const noexcept
+    {
+        return _matrix.cols();
+    }
+
+    //!\copydoc seqan3::detail::debug_matrix::_sequence1
+    sequence1_t const & sequence1() const noexcept
+    {
+        return _sequence1;
+    }
+
+    //!\copydoc seqan3::detail::debug_matrix::_sequence2
+    sequence2_t const & sequence2() const noexcept
+    {
+        return _sequence2;
+    }
+
+    //!\copydoc seqan3::detail::Matrix::at
+    entry_type at(size_t const row, size_t const col) const noexcept
+    {
+        assert(row < rows() && col < cols());
+        return _matrix.at(row, col);
+    }
+
+protected:
+    //!\cond
+    struct format_type; // forward declaration
+    //!\endcond
+
+public:
+    /*!\brief Prints this matrix into the given stream.
+     * \param cout  The stream to print to.
+     * \param flags Modify the way the matrix is printed.
+     *
+     *
+     * The matrix will be printed out with unicode characters if seqan3::fmtflags2::utf8 is set in the flags. Ascii
+     * otherwise.
+     */
+    template <typename oostream_t>
+    void print(oostream_t & cout, fmtflags2 const flags) const noexcept
```

Comment:
why is there a public print function if the debug_matrix should be printed over the debug_stream?

Reply:
debug_matrix is currently in seqan3::detail and no public API.

Altogether, it is more convenient to write the print kernel within a class. 

We could also make the print function protected/ private but then I need to somehow befriend the `operator<<` and I haven't figured out how to do that.